src/lib/yang/xpath.lua

```diff
@@ -0,0 +1,283 @@
+-- Table fields:
+--    This is a table which has a "name" property specifying member
+--    name and has a "keys" (not key) property which is either:
+--       - A string representing the key if the table is string keyed.
+--       - A lua table with corrisponding leaf names as the key and the
+--         value as the value.
+module(..., package.seeall)
+
+local ffi = require("ffi")
+local lib = require("core.lib")
+local schemalib = require("lib.yang.schema")
+local datalib = require("lib.yang.data")
+local valuelib = require("lib.yang.value")
+local util = require("lib.yang.util")
+local normalize_id = datalib.normalize_id
+
+local function extract_parts(fragment)
+   local rtn = {query={}}
+   rtn.name = string.match(fragment, "([^%[]+)")
+   for k,v in string.gmatch(fragment, "%[([^=]+)=([^%]]+)%]") do
+      rtn.query[k] = v
+   end
+   return rtn
+end
+
+local handlers = {}
+function handlers.scalar(fragment, tree)
+   return fragment.name, tree
+end
+function handlers.struct(fragment, tree)
+   return fragment.name, tree.members
+end
+function handlers.table(fragment, tree)
+   return {name=fragment.name, keys=fragment.query}, tree
+end
+function handlers.array(fragment, tree)
+   local position = fragment.query["position()"]
+   return {name=fragment.name, key=tonumber(position)}
+end
+function handle(node_type, fragment, tree)
+   return assert(handlers[node_type], node_type)(fragment, tree)
+end
+
+-- Gets the next item in the path returning the element and the remaining
+-- path fragment. For example "router.routes.route" will return "router"
+-- and "routes.route". If the end is reached it'll return nil.
+local function next_element(path)
+   return string.match(path, "([^/]+)/?(.*)")
+end
+
+local function split_path(path)
+   local tail = path
+   return function ()
+      local head
+      head, tail = next_element(tail)
+      if head == nil then return head else return extract_parts(head) end
+   end
+end
+
+-- Finds the grammar node for a fragment in a given grammar.
+local function extract_grammar_node(grammar, fragment)
+   local errmsg = "Invalid path: "..fragment.name
+   if grammar.type == "table" then
+      if grammar.keys[fragment.name] == nil then
+         return assert(grammar.values[fragment.name], errmsg)
+      else
+         return grammar.keys[fragment.name]
+      end
+   else
+      return assert(grammar[fragment.name], errmsg)
+   end
+end
+
+-- Converts an XPath path to a lua array consisting of path componants.
+-- A path component can then be resolved on a yang data tree:
+function convert_path(grammar, path)
+   local ret = {}
+   local node = grammar
+   for element in split_path(path) do
+      node = extract_grammar_node(node, element)
+      local luapath, next_node = handle(node.type, element, node)
+      table.insert(ret, luapath)
+      node = next_node
+   end
+   return ret
+end
+
+-- Returns a resolver for a paticular schema and *lua* path.
+function resolve(schema, path)
```

Comment:
Consider renaming this to `resolver`, as it just computes the resolver and doesn't actually run the thing